src/full/Agda/Interaction/BasicOps.hs

```diff
@@ -1125,15 +1130,53 @@ introTactic pmLambda ii = do
         makeName ("_", t) = ("x", t)
         makeName (x, t)   = (x, t)
 
-    introData :: I.Type -> TCM [String]
+    introData :: I.Type -> TCM [(String , Maybe A.Expr)]
     introData t = do
       let tel  = telFromList [defaultDom ("_", t)]
           pat  = [defaultArg $ unnamed $ debruijnNamedVar "c" 0]
       r <- splitLast CoInductive tel pat
       case r of
         Left err -> return []
-        Right cov ->
-           mapM showUnambiguousConName $ concatMap (conName . scPats) $ splitClauses cov
+        Right cov -> mapM conWithArgs $ catMaybes $ map (conName . scPats) $ splitClauses cov
+      where
+
+        bindingList :: Telescope1 -> [NamedArg Binder]
+        bindingList =
+          concatMap (\(A.TBind _ _ x _) -> toList x  )
+          . toList
+
+        argsFromTy :: Expr -> [(Maybe String , Hiding)] 
+        argsFromTy = \case
+          A.Pi _ t cd -> (map (\na -> (((render . pretty) <$> (nameOf $ unArg na)) , getHiding na ) ) $ bindingList t ) ++ argsFromTy cd
```

Comment:
just found out that this fails to provide a correct introduction when they are irrefutable patterns in the bindings of Pi.
I plan to address this later today.